plugins/weida-raw/fu-weida-raw-device.c

```diff
@@ -0,0 +1,872 @@
+	fu_progress_step_done(progress);
+
+	fu_device_sleep(device, 2 * 1000);
+
+	if (!fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL)) {
+		fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	} else {
+		fu_wdt_hid_device_rebind_driver(self, error);
+	}
+
+	/* TODO: verify each block */
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_set_quirk_kv(FuDevice *device,
+				 const gchar *key,
+				 const gchar *value,
+				 GError **error)
+{
+	/* TODO: parse value from quirk file */
+	return TRUE;
+	/* failed */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "quirk key not supported");
+	return FALSE;
+}
+
+static void
+fu_weida_raw_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 57, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 43, "reload");
+}
+
+static gchar *
+fu_weida_raw_device_convert_version(FuDevice *device, guint64 version_raw)
+{
+	return fu_version_from_uint32(version_raw, fu_device_get_version_format(device));
+}
+
+static void
+fu_weida_raw_device_init(FuWeidaRawDevice *self)
+{
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_HEX);
+	fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	fu_device_add_protocol(FU_DEVICE(self), "com.weida.raw");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UNSIGNED_PAYLOAD);
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_ONLY_WAIT_FOR_REPLUG);
+	fu_device_add_icon(FU_DEVICE(self), "input-tablet");
+	fu_device_set_name(FU_DEVICE(self), "CoolTouch Device");
+	fu_device_set_vendor(FU_DEVICE(self), "WEIDA");
+	fu_device_set_firmware_gtype(FU_DEVICE(self), FU_TYPE_WEIDA_RAW_FIRMWARE);
+
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_REQUIRE_AC);
+	if (FU_DEVICE(self) != NULL) {
+		const gchar *back_id = fu_device_get_backend_id(FU_DEVICE(self));
+		if (back_id != NULL) {
+			const gchar *i2c_id =
+			    g_strrstr(fu_device_get_backend_id(FU_DEVICE(self)), "i2c");
+			if (i2c_id != NULL)
+				fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL);
+		}
+	}
+	fu_udev_device_add_open_flag(FU_UDEV_DEVICE(self), FU_IO_CHANNEL_OPEN_FLAG_READ);
+	fu_udev_device_add_open_flag(FU_UDEV_DEVICE(self), FU_IO_CHANNEL_OPEN_FLAG_WRITE);
+	fu_udev_device_add_open_flag(FU_UDEV_DEVICE(self), FU_IO_CHANNEL_OPEN_FLAG_NONBLOCK);
+}
+
+static void
+fu_weida_raw_device_finalize(GObject *object)
```

Comment:
we can completely remove fu_weida_raw_device_finalize if there's nothing to unref